keps/sig-node/5304-dra-attributes-downward-api/README.md

```diff
@@ -0,0 +1,830 @@
+
+### Risks and Mitigations
+
+Risk: Exposing device attributes might leak sensitive information.
+Mitigation: Only one attribute is exposed per reference. The NodeAuthorizer ensures kubelet only accesses ResourceClaims and ResourceSlices for Pods scheduled to that node. Attributes originate from the DRA driver via `ResourceSlice`; cluster policy should ensure sensitive data is not recorded there.
+
+Risk: Kubelet performance impact from watching ResourceClaim and ResourceSlice objects.
+Mitigation: Use shared informers with proper indexing to minimize API server load. Scope watches to node-local resources only. Monitor cache memory usage and set reasonable limits.
+
+Risk: API surface expansion could make the Downward API overly complex.
+Mitigation: Keep the API minimal and type-safe with a dedicated selector; no arbitrary JSONPath. Require feature gate enablement in alpha to gather feedback before expanding.
+
+Risk: Compatibility with future DRA changes.
+Mitigation: Implementation is decoupled from DRA driver interfaces. Changes to DRA object structure are handled in the kubelet controller. Attribute names are standardized and versioned.
+
+## Design Details
+
+### API Changes
+
+#### New Downward API Selector: ResourceSliceAttributeSelector
+
+A new typed selector is introduced in `core/v1` that can be used in both environment variable sources and projected downward API volume items:
+
+```go
+// ResourceSliceAttributeSelector selects a DRA-resolved device attribute for a given claim+request.
+// +featureGate=DRADownwardDeviceAttributes
+// +structType=atomic
+type ResourceSliceAttributeSelector struct {
+    // ClaimName must match pod.spec.resourceClaims[].name.
+    // +required
+    ClaimName string `json:"claimName"`
+    
+    // RequestName must match the corresponding ResourceClaim.spec.devices.requests[].name.
+    // +required
+    RequestName string `json:"requestName"`
+    
+    // Attribute specifies which device attribute to expose from the ResourceSlice.
+    // The attribute name must be present in the ResourceSlice's device attributes.
+    // +required
+    Attribute string `json:"attribute"`
+}
+
+// In core/v1 EnvVarSource:
+type EnvVarSource struct {
+    // ...existing fields...
+    
+    // ResourceSliceAttributeRef selects a DRA device attribute for a given claim+request.
+    // +featureGate=DRADownwardDeviceAttributes
+    // +optional
+    ResourceSliceAttributeRef *ResourceSliceAttributeSelector `json:"resourceSliceAttributeRef,omitempty"`
+}
+
+// (Projected volume support deferred to beta)
+```
+
+Validation:
+- Enforce exactly one of `fieldRef`, `resourceFieldRef`, or `resourceSliceAttributeRef` in both env and volume items
+- Validate `claimName` and `requestName` against DNS label rules
+- No API-level enumeration of attribute names; kubelet resolves attributes that exist in the matching `ResourceSlice` at runtime
+
+
+####
+
+### Kubelet Implementation
```

Comment:
do we want the presence of attributes to be a runtime failure as described? Or make it pod admission failure and proactively check the whole Pod's containers ahead of time when pod sandbox is being created?

The bad thing about runtime failure - for pods with restart policy Always -runtime failure will mean that the Pod will get stuck in crash loop backoff

Reply:
it should very much be the odd and exceptional case that the RS is late or gone. Remember, it had to already exist to be selected. I am not worried about races where somehow the driver runs, publishes the resource slice, pod gets scheduled, the kubelet picks it up, but it hasn't yet seen the RS. That seems highly unlikely. The RS informer would have to be way way behind the Pod informer on the same kubelet and apiserver connection, which AFAIK seems unrealistic?

Reply:
@johnbelamaric for the attribute resolution to happen, we need both RS informer and ResourceClaim(RC) informer to be caught up.  While I agree with you that RS slow seems unrealistic(and probably not worth solving for, assuming a lot of other things will fail at that point) but the RC informer which provides the latest RC that was created for this pod could be behind as well, leading to issues I mentioned above.